import time
from PySide6.QtWidgets import QWidget
from PySide6.QtCore import Qt, Signal
# íŒŒì¼ ìœ„ì¹˜ê°€ ê¼¬ì˜€ì„ ìˆ˜ ìˆìœ¼ë‹ˆ ëª…í™•íˆ ì„í¬íŠ¸
try:
    from page_ui.pages.ui_control import Ui_ControlForm
except ImportError:
    from page_ui.ui_control import Ui_ControlForm

class ControlPage(QWidget):
    # MainWindowì™€ ì¼ì¹˜ì‹œí‚¨ ì‹œê·¸ë„: (mode, raw, parsed)
    interactionCommand = Signal(str, object, dict)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.ui = Ui_ControlForm()
        self.ui.setupUi(self)

        self.is_connected = False

        # UI ë¹„ìœ¨ ê°•ì œ ì¡°ì • (ì œì–´íŒ 1 : ì¹´ë©”ë¼ 2)
        self.ui.horizontalLayout_root.setStretch(0, 1)
        self.ui.horizontalLayout_root.setStretch(1, 2)

        self.init_style()
        self.bind_events()

    def init_style(self):
        # ì¹´ë©”ë¼ ë·° ëŒ€ê¸° í™”ë©´ ë° í…Œë‘ë¦¬ í…Œë§ˆ ì ìš©
        self.ui.frameCamera.setStyleSheet("""
            background-color: #000000;
            border: 2px solid #333a45;
            border-radius: 20px;
        """)
        self.ui.lbl_camera_view.setText("WAITING FOR AGV CAMERA...")
        self.ui.lbl_camera_view.setStyleSheet("color: #777; font-weight: bold; border: none;")

        # ë²„íŠ¼ ë° ê·¸ë£¹ë°•ìŠ¤ ìŠ¤íƒ€ì¼
        self.setStyleSheet("""
            QGroupBox { font-size: 15px; font-weight: bold; color: white; border: 1px solid #444; margin-top: 5px; padding-top: 15px; }
            QLabel { font-size: 13px; color: #ddd; }
            QPushButton#btn_connect { background-color: #2d5af1; font-size: 18px; font-weight: bold; color: white; border-radius: 10px; }
            QPushButton { background-color: #2a2f3b; color: white; border-radius: 8px; font-size: 16px; }
            QSlider::handle:horizontal { background: #60a5fa; width: 22px; height: 22px; margin: -8px 0; border-radius: 11px; }
        """)

    def bind_events(self):
        self.ui.btn_connect.clicked.connect(self.toggle_connection)

        # ğŸš€ ìˆ˜ë™ ì´ë™ ë²„íŠ¼ ì´ë²¤íŠ¸ (Pressed=ì „ì†¡, Released=ì •ì§€)
        # ëŒë‹¤ í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•˜ì—¬ ì¸ìë¥¼ ì •í™•íˆ ì „ë‹¬
        self.ui.btn_up.pressed.connect(lambda: self.emit_mqtt("move", "forward"))
        self.ui.btn_down.pressed.connect(lambda: self.emit_mqtt("move", "backward"))
        self.ui.btn_left.pressed.connect(lambda: self.emit_mqtt("move", "left"))
        self.ui.btn_right.pressed.connect(lambda: self.emit_mqtt("move", "right"))

        # ë²„íŠ¼ì—ì„œ ì†ì„ ë–¼ë©´ STOP
        stop_btns = [self.ui.btn_up, self.ui.btn_down, self.ui.btn_left, self.ui.btn_right]
        for b in stop_btns:
            b.released.connect(lambda: self.emit_mqtt("move", "stop"))

        self.ui.btn_stop.clicked.connect(lambda: self.emit_mqtt("move", "stop"))

        # ì†ë„ ë° ê·¸ë©
        self.ui.sld_speed.valueChanged.connect(lambda v: self.emit_mqtt("speed", v))
        self.ui.dial_grab.valueChanged.connect(self._on_grab_changed)

        # ì„œë³´ 1, 2, 3 (4, 5ëŠ” UIì—ì„œ ì‚­ì œë  ì˜ˆì •ì´ë¯€ë¡œ 3ê¹Œì§€ë§Œ ì—°ê²°)
        for i in range(1, 4):
            if hasattr(self.ui, f"sld_s{i}"):
                sld = getattr(self.ui, f"sld_s{i}")
                sld.valueChanged.connect(lambda v, n=i: self._on_servo_changed(n, v))

    def _on_grab_changed(self, v):
        self.ui.lbl_grab_title.setText(f"Grab (S4): {v}Â°")
        self.emit_mqtt("arm", {"servo": 4, "angle": v})

    def _on_servo_changed(self, n, v):
        lbl = getattr(self.ui, f"lbl_s{n}")
        lbl.setText(f"S{n}: {v}Â°")
        self.emit_mqtt("arm", {"servo": n, "angle": v})

    def toggle_connection(self):
        self.is_connected = not self.is_connected
        if self.is_connected:
            self.ui.btn_connect.setText("AGV CONNECTED")
            self.ui.btn_connect.setStyleSheet("background-color: #10b981; color: white; font-weight: bold;")
            self.emit_mqtt("system", "connect")
        else:
            self.ui.btn_connect.setText("CONNECT TO AGV")
            self.ui.btn_connect.setStyleSheet("background-color: #2d5af1; color: white;")
            self.emit_mqtt("system", "disconnect")

    def emit_mqtt(self, action, value):
        # ì—°ê²° ì•ˆ ëì„ ë•Œ system ëª…ë ¹ ì™¸ì—ëŠ” ë¬´ì‹œ
        if not self.is_connected and action != "system":
            return

        # ğŸš€ MainWindow.pyê°€ ì›í•˜ëŠ” 3ë‹¨ êµ¬ì¡° (input_mode, raw_data, parsed_data)
        parsed_data = {
            "mqtt_topic": "interaction/parsed",
            "mqtt_payload": {
                "action": action,
                "value": value,
                "ts": int(time.time() * 1000)
            }
        }
        self.interactionCommand.emit(action, str(value), parsed_data)

    def set_camera_pixmap(self, pixmap):
        if pixmap and not pixmap.isNull():
            self.ui.lbl_camera_view.setPixmap(pixmap.scaled(
                self.ui.lbl_camera_view.size(), Qt.KeepAspectRatio, Qt.SmoothTransformation))
